[WIP] markdown: Fix fenced code bug when message is quote and reply.

CuriousMonkey7 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes part of #17610, #15672
**Testing plan:** <!-- How have you tested? -->
Automated tested using ./tools/test-backend.
Manually tested on all the cases discussed in #17610, #15672, and given on [Common Mark](https://spec.commonmark.org/0.29/#fenced-code-blocks). All cases are showing the expected results after server-side rendering. Frontend part is still showing the wrong results.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

![15672](https://user-images.githubusercontent.com/70478296/111083678-b9630500-8534-11eb-9470-91174091c904.png)

timabbott commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -237,27 +237,60 @@ def done(self) -> None:
         self.processor.pop()
 
 
+class InnerHandler(BaseHandler):
```
So, I'm a bit reluctant about this change, because this file is a forked copy of python-markdown upstream's file of the same name, and rewriting it with a different pattern means it'd be more or less impossible to rebase on top of upstream's bug fixes.  Maybe that's already the case, and/or this isn't significant (because mostly the things you're refactoring are new classes we added ourselves)?   May be worth a look at upstream for this question.

100RABHpy replied:
Yes, it's already the case. We're only using the preprocessor (in FencedBlockPreprocessor) class of Python-markdown. Python markdown doesn't even allow to not close fence, but in Zulip, we're autoclosing fences.

100RABHpy replied:
Yes, I'm only refactoring the classes we already have because there is a lot of repetition in code, and also, to fix #17610 I need all the handler(except the OuterHandler) classes to have one function.

timabbott commented:
@100RABHpy thanks for working on this!  For the frontend markdown processor, a temporary option is to add the test cases but use `marked_execpted_output` for the errors the frontend markdown processor will make in the meantime.

Also, I think it might be better to restructure this PR as first a commit with no functional changes (please run tests) that does the InnerHandler refactoring -- that commit will probably be easier to merge (and will shrink the PR) than the one trying to address the bug.

100RABHpy commented:
Steve Howell also reviewed PR and suggested that I can optimize the code. Right now, I'm working on it. I will restructure PR with the next commit.

timabbott commented on zerver/lib/markdown/fenced_code.py:
```diff
@@ -252,16 +253,29 @@ def __init__(
         else:
             self.lines.append(line.rstrip())
+            self.need_to_handle_unclosed_fence = False
+            return self.need_to_handle_unclosed_fence
```
We definitely don't want to copy this block of logic into 4 different classes -- we should do whatever preparatory refactor commit lets us avoid doing so.

100RABHpy replied:
Okay, I will refactor this. Right now, I just submitted this to show what I'm thinking of a possible solution. I will clean it in future PR.

CuriousMonkey7 committed:
markdown: Refactor repeating code in different Handlers.

In the current implementation of different handlers, there is a lot
of code that is repeating. This commit refactors the code
using some OOP technique to improve code readability.

To refactor this, we first create a class name "InnerHandler" to
clean the handler class’s code (except the OuterHandler class).
This class will be inherited by all the handlers(except the OuterHandler).
Creating this class helps to create class variables in one place.

After creating InnerHandler class, we start refactoring with
"handle_line" function, this function is the same in almost all
handler class. The only difference is the else statement. So,
now we have “handle_line” function in InnerHandler class and
else statement works differently depending on the handler.

After this, we refactor the "done" function. In this function,
last five lines are the same for all the handlers. To refactor this,
we create a new function, "process_line" in Innerhandler
class.

CuriousMonkey7 committed:
markdown: Fix fenced code bug when message is quote and reply.

Fixes server-side part of #15672 and #17610.
In Zulip, it's a feature that if someone opens a fence and
do not close it, the message will be rendered correctly.
But when it's quote_and_reply, it performs unexpectedly.
This commit resolves that error.

To resolve, whenever we encounter a closing fence, we first check the
fence with the handler's fence. If it matches with the handler's
fence, then we close it. If not, we return from the handler and pass
the fence to the previous handler; we do this in "handle_line"
function of FencedBlockPreprocessor.We pass the fence to previous
handler until either the closing fence is matched with an open fence
or there is no match.

In case of a match, we close the fence and pop all the handlers in
the stack present on top of the current handler, we do this
"pop_next_handler" function of FencedBlockPreprocessor. In case of
no match, we render the line normally.

In last, added some tests to verify correct wotrking of code.

gnprice reviewed (commented):
For cross-reference: there's a [chat thread](https://chat.zulip.org/#narrow/stream/9-issues/topic/fenced.20code.20bug/near/1132466) where most of the review of this PR so far has taken place.

(Also I see I wrote an inline comment the other day on one particular piece and didn't hit submit on the review -- so here it is.)
